### PR TITLE
Cleaning up messy implementation

### PR DESCRIPTION
**Pull request by tochman**
PT Story https://www.pivotaltracker.com/story/show/137978291

* introduces `Asset` class
* removes api calls from  `ReportGenerator` module
* refactors `PagesController`

Documented in https://medium.com/craft-academy/cleaning-up-messy-implementation-f8cce0173f66#.mw7w16pb3

### CODE REVIEW

File: app/controllers/pages_controller.rb [hunk 1]
```diff
@@ -44,21 +43,17 @@ def analytics
...
-
-      @image = ReportGenerator.generate(@service, params, @property)
-      message = "#{session[:user_name]} generated a report for #{@property.name}"
+      asset = Asset.new(@service, params)
+      @property = asset.property
+      @image = asset.image
       render :analytics
     rescue Google::Apis::AuthorizationError => ex
       if ex.message
-        render jason: {message: ex}
```
Comment (yggie) [inline app/controllers/pages_controller.rb:55]: 😂

Comment (yggie) [reply app/controllers/pages_controller.rb:55]: on a more serious note, behaviour has changed without a change in tests, this is a good indication that certain tests are missing 😉

File: app/modules/asset.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    @params = params
+    @property = get_property
+    @basic_stats = get_basic_stats
+    @sources = get_sources
+    @os_sources = get_os_sources
+    @traffic_sources = get_traffic_sources
+    @country_sources = get_country_sources
+    @previous_pariod, @current_period = basic_stats.rows.each_slice(basic_stats.rows.size/2).to_a
+    @current_start_date = set_date(current_period[0][0])
+    @current_end_date = set_date(current_period[-1][0])
```
Comment (yggie) [inline app/modules/asset.rb]: in many of these methods, there is an implicit dependency on certain instance variables to be defined. For example, this constructor will break if we move the line:

```ruby
  @service = service
```
to somewhere near the end of the method body. We can make this more explicit and also reduce dependency on the order of evaluation by making methods accept arguments explicitly, for example:

```ruby
@property = get_property

# can be changed to:

@property = get_property(service, params)
```

File: app/modules/asset.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
+class Asset
```
Comment (yggie) [inline app/modules/asset.rb]: I don’t think `Asset` is a great name, it doesn’t describe well the responsibilities of the object. The object looks like it is retrieving the analytics data for a specific time window. This data is then used to generate reports. Now based on my assumptions, I think a name like `AnalyticsWindowReport` would be more appropriate (or anything else, as long as it covers the responsibilities nicely).

File: app/modules/asset.rb [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  def get_property
+    @service.get_web_property(@params[:account_id],
+                              @params[:web_property_id])
+  end
+
+  def get_basic_stats
+    profile_id = "ga:#{@params[:profile_id]}"
+    start_date = (Date.parse(DEFAULT_START_DATE) - 6.days).strftime('%F')
+    end_date = DEFAULT_END_DATE
+    metrics = 'ga:sessions, ga:uniquePageviews'
```
Comment (yggie) [inline app/modules/asset.rb]: a lot of hardcoded `ga: ` strings around, that makes us pretty coupled to Google Analytics. Now that is probably fine, but I just want to check if we are content with that coupling.

File: app/modules/asset.rb [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  def current_page_views
+    summarize_page_views(@current_period)
+  end
+
+  def previous_page_views
+    summarize_page_views(@previous_pariod)
+  end
+
+  def image
+    ReportGenerator.generate(self)
```
Comment (yggie) [inline app/modules/asset.rb]: With this method, the `ReportGenerator` is strongly coupled to the report asset itself. Would it make sense to move this logic higher up into the controller and let the controller decide how the reports will be generated? It feels like this method is mixing two different concerns into the class (most of the class is concerned with the in-memory _representation_ and _collection_ of the data, whereas this method is more about _presentation_)

**Review by yggie — CHANGES_REQUESTED** [review]

File: app/controllers/pages_controller.rb (hunk 1 above)
Comment (tochman) [reply app/controllers/pages_controller.rb:55]: Haha

Comment (tochman) [reply app/controllers/pages_controller.rb:55]: A lot of test are missing. ;-)

File: app/modules/asset.rb (hunk 3 above)
Comment (tochman) [reply app/modules/asset.rb]: I agree. How about `AnalyticsPeriod`? I was considering `Period` (it was called that in my first version, but that also didn't feel right.

File: app/modules/asset.rb (hunk 2 above)
Comment (tochman) [reply app/modules/asset.rb]: I see your point. I considered this approach but decided agains it. Good to get a second opinion and I concur.

File: app/modules/asset.rb (hunk 4 above)
Comment (tochman) [reply app/modules/asset.rb]: At this point we are targeting the GA API so I would not worry about this for now.

**Comment by tochman** [comment]
@yggie && @lollypop27 I’m aware that there are still quite a few possibilities to make the code better. It’s a matter of priorities at this stage in the project. What we are dealing with is a rough prototype so I would not spend too much time on getting every part in to excellent shape.

**Review by yggie — APPROVED** [review]

**Comment by lollypop27** [comment]
@tochman thanks for all this work

**Review by lollypop27 — APPROVED** [review]

**Pull request merged**